### PR TITLE
Add connection pool option

### PR DESCRIPTION
**Pull request by cleaton**
This is a re-opening of PR #330

### CODE REVIEW

File: test/abstractServer.cpp [hunk 1]
```diff
@@ -51,6 +51,11 @@ static void EventHandler(mg_connection* conn, int event, void* event_data) {
 
         case MG_EV_HTTP_REQUEST: {
             AbstractServer* server = static_cast<AbstractServer*>(conn->mgr->user_data);
+            auto port_cstr = (char*) malloc(256);
```
Comment (COM8) [inline test/abstractServer.cpp:54]: Please do not use auto where ever possible.

`malloc` is C, consider replacing it with `std::array`.

**Review by COM8 — COMMENTED** [review]
Thanks for the PR. With a bit of additional work, this will be quiet the nice addition to CPR.

File: test/abstractServer.cpp (hunk 1 above)
Comment (cleaton) [reply test/abstractServer.cpp:54]: The mongoose webserver mg_conn_addr_to_str requires a char* buffer as input as it's written in C. Would it be better I use stack allocation for the char array?

I'll remove auto for the obvious types.

**Comment by cleaton** [comment]
Currently the osx tests are failing with `libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument`. I have not been able to reproduce it on linux so far.

File: test/abstractServer.cpp (hunk 1 above)
Comment (COM8) [reply test/abstractServer.cpp:54]: You can get the underlying `char*` from a `std::array` with the `std::array::data()` call: https://en.cppreference.com/w/cpp/container/array/data

In this case it's better tu use stack memory here sine than we do not have to care about free any more.

**Comment by COM8** [comment]
> Currently the osx tests are failing with `libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument`. I have not been able to reproduce it on linux so far.

Oh, how I love macOS.
I will have a look at it.

**Comment by cleaton** [comment]
> > Currently the osx tests are failing with `libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument`. I have not been able to reproduce it on linux so far.
> 
> Oh, how I love macOS.
> I will have a look at it.

I am wondering if make_shared for the mutex is failing to initialize the underlying mutex implementation?

**Comment by COM8** [comment]
To me it looks like a use after free, but I do not know where.
I will try enable stack traces in gtest so we have a clear indication where this is happening later today.

**Comment by COM8** [comment]
> > > Currently the osx tests are failing with `libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument`. I have not been able to reproduce it on linux so far.
> > 
> > 
> > Oh, how I love macOS.
> > I will have a look at it.
> 
> I am wondering if make_shared for the mutex is failing to initialize the underlying mutex implementation?

This could be a reason.

**Comment by zodinyac** [comment]
Hi, @cleaton! What is current state of this PR? Will you finalize it?

**Comment by cleaton** [comment]
Hi @zodinyac! At the moment I am working on other projects so I won't have time to work on this PR anytime soon. I'll leave it here in case anyone wants to take over and resolve the macos test issue.

**Comment by bokket** [comment]
Hi, @cleaton .Why not try to use the `std::shared_mutex` in c++17 ?
I think this lock is more useful than `std::shared_ptr<std::mutex>` inside a pool of multiple possible connect objects. After all, this is the native synchronization primitive.
https://stackoverflow.com/questions/46452973/difference-between-stdmutex-and-stdshared-mutex

**Comment by COM8** [comment]
Closing this for now since a lot changed since back then.

**Pull request closed without merging**